README.md
# date-format-ms

Formatting of date and times in JS using format codes that match the PHP date() function.

For example:

```js
import { dateToLocalFormat } from 'date-format-ms';

dateToLocalFormat(new Date(), 'D jS M Y g:ia'); // "Monday 9th Aug 2021 5:30pm"
```

-----

Note that for the examples in this document, the system timezone is taken to be Europe/Istanbul, where the local time
is UTC+3:00 (all year).

-----

## Format codes ##

Formatting functions in this package work by supplying a string representing the required format using various codes, as defined in the table below.

We use the same codes as the [PHP date() function](https://www.php.net/manual/en/datetime.format.php) (although not all of the PHP date() codes are supported).

For example

```js
ymdHisToFormat('2021-09-08 17:30:00', 'D jS M Y g:ia'); // "Monday 9th Aug 2021 5:30pm"
```

Unrecognized characters in the format string are written into the output string unchanged, for example:

```js
ymdToFormat('2021-09-08', 'd/M/Y'); // "08/09/2021"
```

To add a literal character to the output you can escape with a backslash, for example:

```js
ymdToFormat('2021-09-08', '\\M: M'); // "M: Aug"
```

| Code | Description | Example |
|------|-------------|---------|
|`d`|Day of the month, 2 digits with leading zeros|`01` to `31`|
|`D`|A textual representation of a day, three letters|`Mon` through `Sun`|
|`j`|Day of the month without leading zeros|`1` to `31`|
|`l`|A full textual representation of the day of the week|`Sunday` through `Saturday`|
|`N`|ISO-8601 numeric representation of the day of the week| `1` (for Monday) through `7` (for Sunday)|
|`w`|Numeric representation of the day of the week|`0` (for Sunday) through `6` (for Saturday)|
|`S`|English ordinal suffix for the day of the month, 2 characters|`st`, `nd`, `rd` or `th`. Works well with `j`|
|`z`|The day of the year (starting from 0)|`0` through `365`|
|`W`|ISO-8601 week number of year, weeks starting on Monday|Example: `42` (the 42nd week in the year)|
|`F`|A full textual representation of a month, such as January or March|`January` through `December`|
|`m`|Numeric representation of a month, with leading zeros|`01` through `12`|
|`M`|A short textual representation of a month, three letters|`Jan` through `Dec`|
|`n`|Numeric representation of a month, without leading zeros|`1` through `12`|
|`t`|Number of days in the given month|`28` through `31`|
|`L`|Whether it's a leap year|`1` if it is a leap year, `0` otherwise.|
|`o`|ISO-8601 week-numbering year. This has the same value as `Y`, except that if the ISO week number (`W`) belongs to the previous or next year, that year is used instead.|Examples: `1999` or `2003`|
|`Y`|A full numeric representation of a year, 4 digits|Examples: `1999` or `2003`|
|`y`|A two digit representation of a year|Examples: `99` or `03`|
|`a`|Lowercase Ante meridiem and Post meridiem|`am` or `pm`|
|`A`|Uppercase Ante meridiem and Post meridiem|`AM` or `PM`|
|`g`|12-hour format of an hour without leading zeros|`1` through `12`|
|`G`|24-hour format of an hour without leading zeros|`0` through `23`|
|`h`|12-hour format of an hour with leading zeros|`01` through `12`|
|`H`|24-hour format of an hour with leading zeros|`00` through `23`|
|`i`|Minutes with leading zeros|`00` to `59`|
|`s`|Seconds with leading zeros|`00` through `59`|
|`v`|Milliseconds.|Example: `654`|
|`U`|Seconds since the Unix Epoch (January 1 1970 00:00:00 GMT), rounded to the nearest second|`1628524558`|

Codes from PHP's date() function that are not supported:

 - Swatch Internet time (`B`) - Not supported because... who needs it?
 - Microseconds (`u`) - Not supported because javascript Date objects only have millisecond precision
 - Various timezone related codes (`e`, `I`, `O`, `P`, `p`, `T`, `Z`) - Not supported because the javascript Date object doesn't contain a timezone, instead it represents a point in time, and your only choices for extracting values from it are UTC or whatever your system's local timezone is configured as.
 - Combination code (`c`,`r`) - Not supported because they contain a timezone indicator (see note above)

-----

## Formatting Functions ##

#### dateToUtcFormat and dateToLocalFormat

`(date: Date, format: string): string`

Takes a javascript Date object and turns it into a string in the supplied format.

If you use `dateToLocalFormat`, the output will be in your local timezone, if you use `dateToUtcFormat` the output will be in the UTC timezone.

```js
import { dateToUtcFormat, dateToLocalFormat } from 'date-format-ms';

const d0 = new Date(2021, 0, 1, 1, 30, 0, 0);
// d0 represents Jan 1st 2021 1:30am in the local (Istanbul) time
dateToLocalFormat(d0, 'Y-m-d H:i:s'); // "2021-01-01 01:30:00"
// However, at this time, in UTC it is still 10:30pm the night before
dateToUtcFormat(d0, 'Y-m-d H:i:s'); // "2020-12-31 22:30:00"

// If you want to specify UTC parameters when creating the JS Date you can do this:
const d1 = new Date(Date.UTC(2021, 0, 1, 1, 30, 0, 0));
// d1 represents Jan 1st 2021 1:30am in UTC
dateToUtcFormat(d1, 'Y-m-d H:i:s'); // "2021-01-01 01:30:00"
// Istanbul is 3 hours ahead
dateToLocalFormat(d1, 'Y-m-d H:i:s'); // "2021-01-01 04:30:00"
```

#### isoToUtcFormat and isoToLocalFormat,

`(iso: any, format: string): string | null`

Takes a datetime represented as a string in ISO8601 format and turns it into a string in the supplied format.

The ISO format consists of a Y-m-d date, followed by the letter 'T' to delimit the time portion, then a time in H:i:s format,
with optional milliseconds, then a timezone indicator - either 'Z' for UTC or an offset like +04:00 or -10:00.

Eg '2021-08-16T12:32:15Z' or '2021-08-16T12:32:15.314845+02:00'

If you use `isoToLocalFormat`, the output will be in your local timezone, if you use `isoToUtcFormat` the output will be in the UTC timezone.

```js
import { isoToUtcFormat, isoToLocalFormat } from 'date-format-ms';

const iso = '2021-08-16T12:30:00-02:00'
// Timezone offset of -02:00, so converting to UTC, this will be:
isoToUtcFormat(iso, 'Y-m-d H:i:s'); // "2021-08-16 14:30:00"
// Istanbul is 3 additional hours ahead
isoToLocalFormat(iso, 'Y-m-d H:i:s'); // "2021-08-16 17:30:00"
```

#### ymdHisToFormat

`(ymdHis: any, format: string): string | null`

Takes a datetime represented as a string in the 'Y-m-d H:i:s' format, and turns it into a string in the supplied format.

If the input is not a string, or is not in the expected 'Y-m-d H:i:s' format, the return value will be `null`.

```js
import { ymdHisToFormat } from 'date-format-ms';

ymdHisToFormat('2021-09-08 10:45:05', 'D jS F g:ia'); // "Monday 9th August 10:45am"
```

#### ymdToFormat

`(ymd: any, format: string): string | null`

Takes a date represented as a string in the 'Y-m-d' format, and turns it into a string in the supplied format.

This is useful for formatting a value that represents just a date, with no time information.

If the input is not a string, or is not in the expected 'Y-m-d' format, the return value will be `null`.

Internally, all time parameters will be zero - IE the instant represented is 00:00:00 on the given date.

```js
import { ymdToFormat } from 'date-format-ms';

ymdToFormat('2021-09-08', 'D jS F'); // "Monday 9th August"
ymdToFormat('2021-09-08', 'D jS F H:i'); // "Monday 9th August 00:00"
```

#### tsToUtcFormat and tsToLocalFormat,

`(ts: any, format: string): string | null`

Takes a javascript timestamp (number) and turns it into a string in the supplied format.

Javascript timestamps are defined as the number of _milliseconds_ since the Unix epoch (1st Jan 1970).
Note that the Unix timestamp is normally defined as the number of _seconds_ since the Unix epoch, so to convert
a Unix timestamp to a Javascript timestamp you can multiply by 1000.

If you use `tsToLocalFormat`, the output will be in your local timezone, if you use `tsToUtcFormat` the output will be in the UTC timezone.

```js
import { tsToUtcFormat, tsToLocalFormat } from 'date-format-ms';

const ts = 1609464600000;
// This is the timestamp for Jan 1st 2021 1:30am in UTC
tsToUtcFormat(ts, 'Y-m-d H:i:s'); // "2021-01-01 01:30:00"
// Istanbul is 3 hours ahead
tsToLocalFormat(ts, 'Y-m-d H:i:s'); // "2021-01-01 04:30:00"
```

-----

## Conversion Functions

#### utcYmdHisToDate and localYmdHisToDate

`(ymdHis: any): Date | null`

Takes a date represented as a string in the 'Y-m-d H:i:s' format, and turns it into a javascript Date object.

If you use `utcYmdHisToDate`, the string will be interpreted as being in UTC.  If you use `localYmdHisToDate` then the
string will be interpreted as being in your local timezone.

If the input is not a string, or is not in the expected 'Y-m-d H:i:s' format, the return value will be `null`.

```js
import { utcYmdHisToDate, localYmdHisToDate } from 'date-format-ms';

const d0 = utcYmdHisToDate('2021-09-08 10:45:05');
d0.getUTCHours(); // 10
d0.getHours(); // 13

const d1 = localYmdHisToDate('2021-09-08 10:45:05');
d1.getUTCHours(); // 7
d1.getHours(); // 10
```

#### utcYmdToDate and localYmdToDate

`(ymd: any): Date | null`

Takes a date represented as a string in the 'Y-m-d' format, and turns it into a javascript Date object.

If you use `utcYmdToDate`, the string will be interpreted as being in UTC (at 00:00).  If you use `localYmdToDate` then the
string will be interpreted as being in your local timezone (at 00:00).

If the input is not a string, or is not in the expected 'Y-m-d' format, the return value will be `null`.

```js
import { utcYmdToDate, localYmdToDate } from 'date-format-ms';

const d0 = utcYmdToDate('2021-09-08');
d0.getUTCHours(); // 0
d0.getHours(); // 3 (local time is 3 hours ahead of UTC)

const d1 = localYmdToDate('2021-09-08');
d1.getUTCHours(); // 21 (UTC is 3 hours behind local time)
d1.getHours(); // 0
```

-----

## Utility Functions

### isLeapYear

`(year: number): boolean`

Returns true if the supplied year is a leap year, false otherwise.

```js
import { isLeapYear } from 'date-format-ms';

isLeapYear(2020); // true
isLeapYear(2021); // false
```

### getNumDaysInMonth

`(date: Date, utc: boolean): number`

Return the number of days in the month for the supplied javascript Date object.

If the second parameter is true UTC month will be used, otherwise the local month will be used.

```js
import { getNumDaysInMonth } from 'date-format-ms';

// 10pm April 30th UTC = 1am May 1st Istanbul (local) time
const d = new Date(Date.UTC(2020, 3, 30, 22));
getNumDaysInMonth(d, true); // 30 - the Date is in April UTC
getNumDaysInMonth(d, false); // 31 - but it's May in local time
```


CHANGELOG.md
# Changelog

## [1.2.0] - 2022-09-22
### Added
 - This changelog file
 - New convenience methods `utcYmdHisToLocalFormat` and `localYmdHisToUtcFormat`


src/index.ts
const shortDayNames: string[] = ['Sun', 'Mon', 'Tue', 'Wed', 'Thu', 'Fri', 'Sat'];
const longDayNames: string[] = ['Sunday', 'Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday'];
const shortMonthNames: string[] = ['Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun', 'Jul', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec'];
const longMonthNames: string[] = [
  'January',
  'February',
  'March',
  'April',
  'May',
  'June',
  'July',
  'August',
  'September',
  'October',
  'November',
  'December',
];
const daysInMonthNotLeapYear: number[] = [31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31];
const microsecondsInDay: number = 1000 * 60 * 60 * 24;
const microsecondsInWeek: number = 1000 * 60 * 60 * 24 * 7;

const numToString = (val: number, withLeadingZeros: boolean): string => {
  const strVal = val.toString(10);
  if (withLeadingZeros) {
    switch (strVal.length) {
      case 0:
        return '00';
      case 1:
        return '0' + strVal;
      default:
        return strVal;
    }
  } else {
    return strVal;
  }
};

const isLeapYear = (year: number): boolean => {
  // Construct a date on the 29th of Feb for the year in question
  // If it gets changed to March 1st then it's not a leap year
  const d1 = new Date(Date.UTC(year, 1, 29));
  return d1.getUTCDate() === 29;
};

const getStartOfDay = (date: Date, utc: boolean): Date => {
  if (utc) {
    return new Date(Date.UTC(date.getUTCFullYear(), date.getUTCMonth(), date.getUTCDate()));
  } else {
    return new Date(date.getFullYear(), date.getMonth(), date.getDate());
  }
};

const getStartOfYear = (date: Date, utc: boolean): Date => {
  if (utc) {
    return new Date(Date.UTC(date.getUTCFullYear(), 0, 1));
  } else {
    return new Date(date.getFullYear(), 0, 1);
  }
};

const getFirstThursdayOfYear = (year: number, utc: boolean): Date => {
  const d = utc ? new Date(Date.UTC(year, 0, 1)) : new Date(year, 0, 1);
  const n = utc ? d.getUTCDay() : d.getDay();
  d.setDate(1 + ((11 - n) % 7));
  return d;
};

const getThursdayOfWeek = (date: Date, utc: boolean): Date => {
  const d: Date = getStartOfDay(date, utc);
  const n = utc ? d.getUTCDay() : d.getDay();
  const offset = ((7 - n) % 7) - 3;
  d.setDate((utc ? d.getUTCDate() : d.getDate()) + offset);
  return d;
};

const getDayOfMonth = (date: Date, utc: boolean): number => {
  // Day of the month without leading zeros
  // 1 to 31
  return utc ? date.getUTCDate() : date.getDate();
};

const getDayOfWeekIso8601 = (date: Date, utc: boolean): number => {
  // ISO-8601 numeric representation of the day of the week
  // 1 (for Monday) through 7 (for Sunday)
  const n = utc ? date.getUTCDay() : date.getDay();
  return n === 0 ? 7 : n;
};

const getDayOfWeek = (date: Date, utc: boolean): number => {
  // Numeric representation of the day of the week
  // 0 (for Sunday) through 6 (for Saturday)
  return utc ? date.getUTCDay() : date.getDay();
};

const getDayOfYear = (date: Date, utc: boolean): number => {
  // The day of the year (starting from 0)
  // 0 through 365
  const tsAtStartOfDay = getStartOfDay(date, utc).getTime();
  const tsAtStartOfYear = getStartOfYear(date, utc).getTime();
  return Math.round((tsAtStartOfDay - tsAtStartOfYear) / microsecondsInDay);
};

const getWeekOfYear = (date: Date, utc: boolean): number => {
  // ISO-8601 week number of year, weeks starting on Monday
  // See https://en.wikipedia.org/wiki/ISO_week_date for explanation of this
  const thisThur = getThursdayOfWeek(date, utc);
  const firstThur = getFirstThursdayOfYear(utc ? thisThur.getUTCFullYear() : thisThur.getFullYear(), utc);
  return 1 + Math.round((thisThur.getTime() - firstThur.getTime()) / microsecondsInWeek);
};

const getMonth = (date: Date, utc: boolean): number => {
  // Numeric representation of a month, without leading zeros
  // 1 through 12
  return (utc ? date.getUTCMonth() : date.getMonth()) + 1;
};

const getYear = (date: Date, utc: boolean): number => {
  // A full numeric representation of a year, 4 digits
  // Examples: 1999 or 2003
  return utc ? date.getUTCFullYear() : date.getFullYear();
};

const getNumDaysInMonth = (date: Date, utc: boolean): number => {
  // Number of days in the given month
  // 28 through 31
  const n = utc ? date.getUTCMonth() : date.getMonth();
  if (n === 1 && isLeapYear(utc ? date.getUTCFullYear() : date.getFullYear())) {
    // February in leap year:
    return 29;
  }
  return daysInMonthNotLeapYear[n];
};

const getWeekNumberingYearIso8601 = (date: Date, utc: boolean): number => {
  // ISO-8601 week-numbering year. This has the same value as Y, except that if the ISO week number (W) belongs to the previous or next year, that year is used instead.
  // Examples: 1999 or 2003
  const thisThur = getThursdayOfWeek(date, utc);
  return utc ? thisThur.getUTCFullYear() : thisThur.getFullYear();
};

const getHour24 = (date: Date, utc: boolean): number => {
  // 24-hour format of an hour
  // 0 through 23
  return utc ? date.getUTCHours() : date.getHours();
};

const getHour12 = (date: Date, utc: boolean): number => {
  // 12-hour format of an hour
  // 1 through 12
  const n = getHour24(date, utc) % 12;
  return n === 0 ? 12 : n;
};

const getMinutes = (date: Date, utc: boolean): number => {
  return utc ? date.getUTCMinutes() : date.getMinutes();
};

const getSeconds = (date: Date, utc: boolean): number => {
  return utc ? date.getUTCSeconds() : date.getSeconds();
};

const getMilliseconds = (date: Date, utc: boolean): number => {
  return utc ? date.getUTCMilliseconds() : date.getMilliseconds();
};

const getOrdinalSuffix = (dayOfMonth: number): string => {
  // English ordinal suffix for the day of the month, 2 characters
  // st, nd, rd or th. Works well with j
  if (dayOfMonth > 3 && dayOfMonth < 21) {
    return 'th';
  }
  switch (dayOfMonth % 10) {
    case 1:
      return 'st';
    case 2:
      return 'nd';
    case 3:
      return 'rd';
  }
  return 'th';
};

const dateTimeChar = (char: string, date: Date, utc: boolean): string => {
  switch (char) {
    case 'd':
      // Day of the month, 2 digits with leading zeros
      // 01 to 31
      return numToString(getDayOfMonth(date, utc), true);
    case 'D':
      // A textual representation of a day, three letters
      // Mon through Sun
      return shortDayNames[utc ? date.getUTCDay() : date.getDay()];
    case 'j':
      // Day of the month without leading zeros
      // 1 to 31
      return numToString(getDayOfMonth(date, utc), false);
    case 'l':
      // A full textual representation of the day of the week
      // Sunday through Saturday
      return longDayNames[utc ? date.getUTCDay() : date.getDay()];
    case 'N':
      // ISO-8601 numeric representation of the day of the week
      // 1 (for Monday) through 7 (for Sunday)
      return numToString(getDayOfWeekIso8601(date, utc), false);
    case 'S':
      // English ordinal suffix for the day of the month, 2 characters
      // st, nd, rd or th. Works well with j
      return getOrdinalSuffix(getDayOfMonth(date, utc));
    case 'w':
      // Numeric representation of the day of the week
      // 0 (for Sunday) through 6 (for Saturday)
      return numToString(getDayOfWeek(date, utc), false);
    case 'z':
      // The day of the year (starting from 0)
      // 0 through 365
      return numToString(getDayOfYear(date, utc), false);
    case 'W':
      // ISO-8601 week number of year, weeks starting on Monday
      // See https://en.wikipedia.org/wiki/ISO_week_date for explanation of this
      return numToString(getWeekOfYear(date, utc), true);
    case 'F':
      // A full textual representation of a month
      // January through December
      return longMonthNames[utc ? date.getUTCMonth() : date.getMonth()];
    case 'm':
      // Numeric representation of a month, with leading zeros
      // 01 through 12
      return numToString(getMonth(date, utc), true);
    case 'M':
      // A short textual representation of a month, three letters
      // Jan through Dec
      return shortMonthNames[utc ? date.getUTCMonth() : date.getMonth()];
    case 'n':
      // Numeric representation of a month, without leading zeros
      // 1 through 12
      return numToString(getMonth(date, utc), false);
    case 't':
      // Number of days in the given month
      // 28 through 31
      return numToString(getNumDaysInMonth(date, utc), false);
    case 'L':
      // Whether it's a leap year
      // 1 if it is a leap year, 0 otherwise.
      return isLeapYear(getYear(date, utc)) ? '1' : '0';
    case 'o':
      // ISO-8601 week-numbering year. This has the same value as Y, except that if the ISO week number (W) belongs to the previous or next year, that year is used instead.
      // Examples: 1999 or 2003
      return numToString(getWeekNumberingYearIso8601(date, utc), false);
    case 'Y':
      // A full numeric representation of a year, 4 digits
      // Examples: 1999 or 2003
      return numToString(getYear(date, utc), false);
    case 'y':
      // A two digit representation of a year
      // Examples: 99 or 03
      return numToString(getYear(date, utc), false).substr(2, 2);
    case 'a':
      // Lowercase Ante meridiem and Post meridiem
      // am or pm
      return getHour24(date, utc) < 12 ? 'am' : 'pm';
    case 'A':
      // Uppercase Ante meridiem and Post meridiem
      // AM or PM
      return getHour24(date, utc) < 12 ? 'AM' : 'PM';
    case 'g':
      // 12-hour format of an hour without leading zeros
      // 1 through 12
      return numToString(getHour12(date, utc), false);
    case 'G':
      // 24-hour format of an hour without leading zeros
      // 0 through 23
      return numToString(getHour24(date, utc), false);
    case 'h':
      // 12-hour format of an hour with leading zeros
      // 01 through 12
      return numToString(getHour12(date, utc), true);
    case 'H':
      // 24-hour format of an hour with leading zeros
      // 00 through 23
      return numToString(getHour24(date, utc), true);
    case 'i':
      // Minutes with leading zeros
      // 00 to 59
      return numToString(getMinutes(date, utc), true);
    case 's':
      // Seconds with leading zeros
      // 00 through 59
      return numToString(getSeconds(date, utc), true);
    case 'v':
      // Milliseconds
      return numToString(getMilliseconds(date, utc), false);
    case 'U':
      // Unix timestamp - seconds since the Unix Epoch (January 1 1970 00:00:00 GMT)
      return numToString(Math.round(date.getTime() / 1000), false);
    default:
      return char;
  }
};

const dateToFormat = (date: Date, format: string, utc: boolean): string => {
  let char: string;
  let out: string = '';
  for (let i = 0; i < format.length; i++) {
    char = format[i];
    if (char === '\\') {
      // Backslash escapes the next character
      i++;
      out += format[i];
    } else {
      out += dateTimeChar(char, date, utc);
    }
  }
  return out;
};

const dateToUtcFormat = (date: Date, format: string): string => {
  return dateToFormat(date, format, true);
};

const dateToLocalFormat = (date: Date, format: string): string => {
  return dateToFormat(date, format, false);
};

const ymdToDate = (ymd: any, utc: boolean): Date | null => {
  if (typeof ymd === 'string') {
    const match = /^(\d\d\d\d)-(\d\d)-(\d\d)/.exec(ymd.replace(/\s/g, ''));
    if (match) {
      const year = parseInt(match[1], 10);
      const monthIndex = parseInt(match[2], 10) - 1;
      const dayOfMonth = parseInt(match[3], 10);
      if (utc) {
        return new Date(Date.UTC(year, monthIndex, dayOfMonth, 0, 0, 0, 0));
      } else {
        return new Date(year, monthIndex, dayOfMonth, 0, 0, 0, 0);
      }
    }
  }
  return null;
};

const utcYmdToDate = (ymd: any): Date | null => {
  return ymdToDate(ymd, true);
};

const localYmdToDate = (ymd: any): Date | null => {
  return ymdToDate(ymd, false);
};

const ymdToFormat = (ymd: any, format: string): string | null => {
  const date = ymdToDate(ymd, true);
  if (date) {
    return dateToFormat(date, format, true);
  } else {
    return null;
  }
};

const ymdHisToDate = (ymdHis: any, utc: boolean): Date | null => {
  if (typeof ymdHis === 'string') {
    // look for Y-m-d H:i:s format (eg 2020-06-30 07:36:15)
    const match = /^(\d\d\d\d)-(\d\d)-(\d\d) (\d\d):(\d\d):(\d\d)$/.exec(ymdHis.trim());
    if (!match) {
      return null;
    }
    const year = parseInt(match[1], 10);
    const monthIndex = parseInt(match[2], 10) - 1;
    const dayOfMonth = parseInt(match[3], 10);
    const hours = parseInt(match[4], 10);
    const mins = parseInt(match[5], 10);
    const secs = parseInt(match[6], 10);
    if (utc) {
      return new Date(Date.UTC(year, monthIndex, dayOfMonth, hours, mins, secs, 0));
    } else {
      return new Date(year, monthIndex, dayOfMonth, hours, mins, secs, 0);
    }
  } else {
    return null;
  }
};

const utcYmdHisToDate = (ymdHis: any): Date | null => {
  return ymdHisToDate(ymdHis, true);
};

const localYmdHisToDate = (ymdHis: any): Date | null => {
  return ymdHisToDate(ymdHis, false);
};

const ymdHisToFormat = (ymdHis: any, format: string): string | null => {
  const date = utcYmdHisToDate(ymdHis);
  if (date) {
    return dateToUtcFormat(date, format);
  } else {
    return null;
  }
};

const utcYmdHisToLocalFormat = (ymdHis: any, format: string): string | null => {
  const date = utcYmdHisToDate(ymdHis);
  if (date) {
    return dateToLocalFormat(date, format);
  } else {
    return null;
  }
};

const localYmdHisToUtcFormat = (ymdHis: any, format: string): string | null => {
  const date = localYmdHisToDate(ymdHis);
  if (date) {
    return dateToUtcFormat(date, format);
  } else {
    return null;
  }
};

const isoToDate = (iso: any): Date | null => {
  if (typeof iso === 'string') {
    // look for iso format eg
    // 2020-06-30T07:36:15+02:00
    // 2019-11-26T13:47:23.342983Z
    const match = /^(\d\d\d\d)-(\d\d)-(\d\d)T(\d\d):(\d\d):(\d\d)(\.\d\d\d\d\d\d)?(([+-])(\d\d):(\d\d)|Z)$/.exec(
      iso.trim(),
    );
    if (!match) {
      return null;
    }
    const year = parseInt(match[1], 10);
    const monthIndex = parseInt(match[2], 10) - 1;
    const dayOfMonth = parseInt(match[3], 10);
    const hours = parseInt(match[4], 10);
    let mins = parseInt(match[5], 10);
    const secs = parseInt(match[6], 10);
    const millisecs = Math.round(match[7] == null ? 0 : parseFloat(match[7]) * 1000);
    if (match[8] !== 'Z') {
      mins += (match[9] === '+' ? -1 : 1) * (parseInt(match[10], 10) * 60 + parseInt(match[11], 10));
    }
    return new Date(Date.UTC(year, monthIndex, dayOfMonth, hours, mins, secs, millisecs));
  } else {
    return null;
  }
};

const isoToFormat = (iso: any, format: string, utc: boolean): string | null => {
  const date = isoToDate(iso);
  if (date) {
    return dateToFormat(date, format, utc);
  } else {
    return null;
  }
};

const isoToUtcFormat = (iso: any, format: string): string | null => {
  return isoToFormat(iso, format, true);
};

const isoToLocalFormat = (iso: any, format: string): string | null => {
  return isoToFormat(iso, format, false);
};

const tsToFormat = (ts: any, format: string, utc: boolean): string | null => {
  if (typeof ts === 'number') {
    const date = new Date(ts);
    return dateToFormat(date, format, utc);
  } else {
    return null;
  }
};

const tsToUtcFormat = (ts: any, format: string): string | null => {
  return tsToFormat(ts, format, true);
};

const tsToLocalFormat = (ts: any, format: string): string | null => {
  return tsToFormat(ts, format, false);
};

export {
  isLeapYear,
  getNumDaysInMonth,
  dateToUtcFormat,
  dateToLocalFormat,
  utcYmdToDate,
  localYmdToDate,
  ymdToFormat,
  utcYmdHisToDate,
  localYmdHisToDate,
  ymdHisToFormat,
  utcYmdHisToLocalFormat,
  localYmdHisToUtcFormat,
  isoToDate,
  isoToUtcFormat,
  isoToLocalFormat,
  tsToUtcFormat,
  tsToLocalFormat,
};
